plugins/Module.h

```diff
@@ -0,0 +1,52 @@
+// Copyright 2020 The Mumble Developers. All rights reserved.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file at the root of the
+// Mumble source tree or at <https://www.mumble.info/LICENSE>.
+
+#ifndef MODULE_H
+#define MODULE_H
+
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+typedef uint64_t procptr_t;
+
+struct MemoryRegion {
+	procptr_t address;
+	size_t size;
+
+	bool readable;
+	bool writable;
+	bool executable;
+
+	bool operator<(const MemoryRegion &region) const { return address < region.address; }
+
+	MemoryRegion() : address(0), size(0), readable(false), writable(false), executable(false) {}
+};
+
+typedef std::vector< MemoryRegion > MemoryRegions;
+
+class Module {
+protected:
+	std::string m_name;
+	MemoryRegions m_regions;
+
+public:
+	inline std::string name() const { return m_name; }
+	inline MemoryRegions regions() const { return m_regions; }
+
+	procptr_t baseAddress() const;
+	MemoryRegions readableRegions() const;
+
+	void addRegion(const MemoryRegion &region);
+
+	Module(const std::string &name);
+};
+
+class Modules : public std::unordered_map< std::string, Module > {
```

Comment:
Does `unordered_map` have a virtual destructor? Have a look at https://stackoverflow.com/questions/2597116/is-extending-a-base-class-with-non-virtual-destructor-dangerous

Reply:
Good catch.

https://stackoverflow.com/a/10477849